modules/adagioRtdProvider.js

```diff
@@ -0,0 +1,690 @@
+
+  loadAdagioScript(config);
+
+  return true;
+}
+
+/**
+ * onBidRequest is called for each bidder during an auction and contains the bids for that bidder.
+ *
+ * @param {*} bidderRequest
+ * @param {*} config
+ * @param {*} _userConsent
+ */
+function onBidRequest(bidderRequest, config, _userConsent) {
+  // setTimeout trick to ensure that the `bidderRequest.params` values updated by a bidder adapter are taken into account.
+  // @todo: Check why we have to do it like this, and if there is a better way. Check how the event is dispatched in rtdModule/index.js
+  setTimeout(() => {
+    bidderRequest.bids.forEach(bid => {
+      const uid = deepAccess(bid, 'ortb2.site.ext.data.adg_rtd.uid');
+      if (!uid) {
+        logError('The `uid` is required to store the request in the ADAGIO namespace.');
+        return;
+      }
+
+      // No need to store the same info several times.
+      // `uid` is unique as it is generated by the RTD module itself for each auction.
+      const key = `${bid.adUnitCode}-${uid}`;
+      if (_internal.getGuard().has(key)) {
+        return;
+      }
+
+      _internal.getGuard().add(key);
+      storeRequestInAdagioNS(bid, config);
+    });
+  }, 1);
+}
+
+/**
+ * onGetBidRequestData is called once per auction.
+ * Update both the `ortb2Fragments` and `ortb2Imp` objects with features computed for Adagio.
+ *
+ * @param {*} bidReqConfig
+ * @param {*} callback
+ * @param {*} config
+ */
+function onGetBidRequestData(bidReqConfig, callback, config) {
+  const { site: ortb2Site } = bidReqConfig.ortb2Fragments.global;
+  const features = _internal.getFeatures().get();
+  const ext = {
+    uid: generateUUID(),
+    features: { ...features },
+    session: { ..._SESSION.get() }
+  };
+
+  deepSetValue(ortb2Site, `ext.data.adg_rtd`, ext);
+
+  const adUnits = bidReqConfig.adUnits || getGlobal().adUnits || [];
+  adUnits.forEach(adUnit => {
+    const ortb2Imp = deepAccess(adUnit, 'ortb2Imp');
+    // A divId is required to compute the slot position and later to track viewability.
+    // If nothing has been explicitly set, we try to get the divId from the GPT slot and fallback to the adUnit code in last resort.
+    if (!deepAccess(ortb2Imp, 'ext.data.divId')) {
+      const divId = getGptSlotInfoForAdUnitCode(adUnit.code).divId;
+      deepSetValue(ortb2Imp, `ext.data.divId`, divId || adUnit.code);
+    }
+
+    const slotPosition = getSlotPosition(adUnit);
+    deepSetValue(ortb2Imp, `ext.data.adg_rtd.adunit_position`, slotPosition);
+
+    // We expect `pagetype` `category` are defined in FPD `ortb2.site.ext.data` object.
+    // `placement` is expected in FPD `adUnits[].ortb2Imp.ext.data` object. (Please note that this `placement` is not related to the oRTB video property.)
+    // Btw, we have to ensure compatibility with publishers that use the "legacy" adagio params at the adUnit.params level.
+    const adagioBid = adUnit.bids.find(bid => _internal.isAdagioBidder(bid.bidder));
+    if (adagioBid) {
```

Comment:
hI @osazos ! I noticed a new commit, and also this special handling for the adagio bidder in this block remains. in https://github.com/prebid/Prebid.js/issues/8596 for Prebid 8, all special handling of rtd information was moved from the rtd module into the bid module. If your bidder needs to look at the injected rtd info differently than a typical bidder, why not put that in your bidder js?

Reply:
Hi @patmmccann,

Let me know if I misunderstand your comment, because the fallback is to handle the opposite: we set for "all" what was set for the adagio adapter.

The case is, when a client has its ad-units already configured at the `adUnits[].adagio.params` level and switch to the rtd Provider, he must move some of these params at the FPD level. In the real life, our clients won't have the flexibility to easily do the switch. Applying this fallback allow us to progressively onboard them.